docs/source/model/dependencies.rst

```diff
@@ -0,0 +1,453 @@
+
+    import mlflow
+
+
+    class MyModel(mlflow.pyfunc.PythonModel):
+        def predict(self, context, model_input):
+            from utils import my_func
+
+            x = my_func(model_input)
+            # .. your prediction logic
+            return prediction
+
+
+    # Log the model
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model(
+            python_model=MyModel(),
+            artifact_path="model",
+            code_paths=["utils.py"],
+        )
+
+Then MLflow will save ``utils.py`` under ``code/`` directory in the model directory:
+
+::
+
+    model/
+    ├── MLmodel
+    ├── ...
+    └── code/
+        └── utils.py
+
+When MLflow loads the model for serving, the ``code`` directory will be added to the system path so that you can use the module in your model
+code like ``from utils import my_func``. You can also specify a directory path as ``code_path`` to save multiple files under the directory:
+
+Caveats of ``code_path`` Option
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When using the ``code_path`` option, please be aware of the limitation that the specified file or directory **must be in the same directory as your model script**.
+If the specified file or direcotry is in a parent or child directory like ``my_project/src/utils.py``, model serving will fail with ``ModuleNotFoundError``.
+For example, let's say if you have the following file structure in your project
+
+::
+
+    my_project/
+    |── train.py
+    └── src/
+        └──  utils.py
+
+Then the following model code does **not** work:
+
+.. code-block:: python
+
+    class MyModel(mlflow.pyfunc.PythonModel):
+        def predict(self, context, model_input):
+            from src.utils import my_func
+
+            # .. your prediction logic
+            return prediction
+
+
+    with mlflow.start_run() as run:
+        mlflow.pyfunc.log_model(
+            python_model=MyModel(),
+            artifact_path="model",
+            code_paths=[
+                "src/utils.py"
+            ],  # the file will be saved at code/utils.py not code/src/utils.py
+        )
+
+    # => Model serving will fail with ModuleNotFoundError: No module named 'src'
+
+This limitation is due to how MLflow saves and loads the specified files/directories. When it copies the specified files or directries in ``code/`` target,
+it does **not** preserve the relative path for them. For instance, in the above example, MLflow will copy ``utils.py`` to ``code/utils.py``, not
+``code/src/utils.py``. As a result, it has to be imported as ``from utils import my_func``, instead of ``from src.utils import my_func``.
+However, this may not be pleasant as the import path is different from the original training script.
+
+One workaround for this issue is to use the parent directory instead, which means doing ``code_path=["src"]`` in this example.
+This way, MLflow will copy the entire ``src/`` directory under ``code/`` and your model code will be able to import ``src.utils``.
+
```

Comment:
Can we add a note for recommended file structures for dependent code inclusion? (we recommend including all dependent files in their own folder that is accessible from the main entry point of your training code, as a directory, that contains no other files or code structures that are not absolutely required for the proper functionality of your model) or something to that effect?

Reply:
Sure, I think what you mentioned is the recommended (or perhaps only viable) way for managing dependent code.